### PR TITLE
Real set variable

### LINKED ISSUE #237: SET_VARIABLE is almost just DEFINE_VARIABLE
I also just realize that our `SET_VARIABLE` is almost just `DEFINE_VARIABLE`,
in script is only possible to do `x = 2` and `x = 3`, but can't do `x = 2 + 1` or `x += 1` or logic in it `x = a or b` - this is bad!
It means that now for base math and logic user need to add custom regex and it shouldn't be this way.

### PR DESCRIPTION
**Pull request by Jeremi360**
fixes #237

### CODE REVIEW

File: addons/Rakugo/lib/systems/Parser.gd [hunk 1]
```diff
@@ -65,7 +65,8 @@ var parser_regex :={
 	# jump label
 	JUMP = "^jump (?<label>{NAME})( if (?<expression>.+))?$",
 	# for setting Rakugo variables
-	SET_VARIABLE = "^(?<lvar_name>{VARIABLE}) = ((?<text>{STRING})|(?<number>{NUMERIC})|(?<rvar_name>{VARIABLE}))$",
+	ASSIGNMENT = "^(=|\\+=|\\-=|\\*=|\\/=|\\%=)",
```
Comment (theludovyc) [inline addons/Rakugo/lib/systems/Parser.gd]: Please move it in var Regex. var parser_regex is only for RkScript instructions.

File: addons/Rakugo/lib/systems/Executer.gd [hunk 2]
```diff
@@ -246,29 +244,55 @@ func do_execute_script(parameters:Dictionary):
...
 
-					if value.is_valid_int():
-						value = int(value)
-					else:
-						value = float(value)
-
-				Rakugo.set_variable(result.get_string("lvar_name"), value)
+				else:
+					Rakugo.set_variable(rvar_name, final_value)
+
```
Comment (theludovyc) [inline addons/Rakugo/lib/systems/Executer.gd]: You do not need if/else here, just a if.

You can do
```gd
if operator != "=":
  #...
  match:
    "+=":
      final_value += org_value
#endif
Rakugo.set_variable(rvar_name, final_value)
```

**Review by theludovyc — CHANGES_REQUESTED** [review]
Do not forget to update unit tests or add new like you say

Empty lines are here to make code more readable, why you remove them ?

**Comment by Jeremi360** [comment]
> Do not forget to update unit tests or add new like you say

I'm doing it, this is way I made this PR a draft in the first place.

> Empty lines are here to make code more readable, why you remove them ?

Wow sorry dude, but you made new line so often that for me it looked bad, so I thought that some thing was wrong.
For me it was harder to read it. I thought it was some formatter or code prettyfier go crazy.
I will try to restore your empty lines.

File: addons/Rakugo/lib/systems/Parser.gd (hunk 1 above)
Comment (Jeremi360) [reply addons/Rakugo/lib/systems/Parser.gd]: Okey I will move it, I didn't know you had such assumptions in the code.

File: addons/Rakugo/lib/systems/Executer.gd (hunk 2 above)
Comment (Jeremi360) [reply addons/Rakugo/lib/systems/Executer.gd]: You are right I will remove else

**Comment by Jeremi360** [comment]
I added tests, but my code doesn't work and I don't know way.
Can you help me @theludovyc ?

**Comment by theludovyc** [comment]
I not forgot you @Jeremi360 😊

**Comment by Jeremi360** [comment]
> Do not forget to update unit tests or add new like you say
> 
> Empty lines are here to make code more readable, why you remove them ?

I think I already restored your empty lines @theludovyc

**Comment by theludovyc** [comment]
@Jeremi360 I mean, to help you 😊

File: addons/Rakugo/lib/systems/Executer.gd [hunk 3]
```diff
@@ -152,7 +150,9 @@ func do_execute_script(parameters:Dictionary):
 					var values = []
 
 					for var_name in line[3]:
-						var var_ = Rakugo.get_variable(var_name)
+						var var_ = null
+						if Rakugo.has_variable(var_name):
+							var_ = Rakugo.get_variable(var_name)
```
Comment (theludovyc) [inline addons/Rakugo/lib/systems/Executer.gd:155]: Rakugo.get_variable return null if variable is not found. So you do not need to use has_variable before get_variable

File: addons/Rakugo/lib/systems/Parser.gd [hunk 4]
```diff
@@ -203,42 +223,33 @@ func parse_script(lines:PackedStringArray) -> Dictionary:
...
-								var_name_expr = sub_result.get_string("var_name")
-							
-							if !vars_expression.has(var_name_expr):
-								vars_expression.push_back(var_name_expr)
+					"SET_VARIABLE":
+						var str_expression:String = result.get_string("expression")
 
+						if str_expression.is_empty():
+							parse_array.push_back([key, result])
+							break
```
Comment (theludovyc) [inline addons/Rakugo/lib/systems/Parser.gd]: If your str_expression is empty why you do not send an error ?

**Review by theludovyc — CHANGES_REQUESTED** [review]

**Comment by theludovyc** [comment]
> I added tests, but my code doesn't work and I don't know way. Can you help me @theludovyc ?

image

@Jeremi360 Do you try to debug with prints ?

**Comment by theludovyc** [comment]
@Jeremi360 You need to fix this bug https://github.com/rakugoteam/Rakugo-Dialogue-System/issues/99, to fix yours

File: addons/Rakugo/lib/systems/Executer.gd (hunk 3 above)
Comment (Jeremi360) [reply addons/Rakugo/lib/systems/Executer.gd:155]: Yes, I made this if this way as Rakugo would crash when `get_variable()` returns null - we program it this way,
`push_error()` stops program no matter what is next.

https://github.com/rakugoteam/Rakugo-Dialogue-System/blob/d11e1fe73a61808a65c9c07bc4194303cae77d95/addons/Rakugo/Rakugo.gd#L74-L85

File: addons/Rakugo/lib/systems/Parser.gd (hunk 4 above)
Comment (Jeremi360) [reply addons/Rakugo/lib/systems/Parser.gd:240]: Yes you are right, I'made a mistakie it should be `if not` and else it should push an error - I fixing it now.

Comment (Jeremi360) [reply addons/Rakugo/lib/systems/Parser.gd]: I'm not sure, I last touch this code over a month ago, so maybe I was to change into something and forgot.
I will test with out this `if`-block and only if find that is really need restore it in proper form.

File: addons/Rakugo/lib/systems/Executer.gd (hunk 2 above)
Comment (Jeremi360) [reply addons/Rakugo/lib/systems/Executer.gd]: Yes, it should work I will try that.

File: addons/Rakugo/lib/systems/Parser.gd (hunk 1 above)
Comment (Jeremi360) [reply addons/Rakugo/lib/systems/Parser.gd]: Is done along a go.

File: addons/Rakugo/lib/systems/Executer.gd (hunk 2 above)
Comment (Jeremi360) [reply addons/Rakugo/lib/systems/Executer.gd]: Is done along a go.

**Comment by Jeremi360** [comment]
@theludovyc 
About https://github.com/rakugoteam/Rakugo-Dialogue-System/pull/240#issuecomment-1669652167 :
I tried to debug it with unit-test and also with debug with prints and like on your screenshot it only catches strings.

About #99 I will try to fix it also then.

When I try to run any of RKScripts examples in Godot 4.1.1 it always start for this 3 errors:
image
So which should we use instead `call_deferred()` or `call_thread_group()` ?

**Comment by theludovyc** [comment]
I found why we have these errors now https://github.com/godotengine/godot/issues/78214#issuecomment-1590802452

Edit : Fix is ready, I made a pr soon :) !

Edit : here https://github.com/rakugoteam/Rakugo-Dialogue-System/pull/241

**Comment by theludovyc** [comment]
Fix is ready, I made a pr soon :) !

**Comment by Jeremi360** [comment]
> I found why we have these errors now [godotengine/godot#78214 (comment)](https://github.com/godotengine/godot/issues/78214#issuecomment-1590802452)
> 
> Edit : Fix is ready, I made a pr soon :) !
> 
> Edit : here #241

@theludovyc Nice! I will merge it soon!

**Comment by Jeremi360** [comment]
partialy implemented by #263

**Pull request closed without merging**